### PR TITLE
sopel: drop support for Python 3.6 and require 3.7+

### PR DESCRIPTION
**Pull request by Exirel**
### Description

Tin. Python 3.6 died yesterday. I checked the warning for 3.6, they were written for Sopel 8, so we can either keep 3.6 or drop it now. Either way, this is not a big change, unlike Python 2.

I haven't modified the usage of `OrderedDict`, because there are slight difference between `dict` and `OrderedDict` in particular when considering equality. 

I haven't check the "future" import either, hence the "draft" status for now.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

**Comment by dgw** [comment]
Might as well take this opportunity to add 3.10 to CI, no?

**Comment by Exirel** [comment]
>  Test suite (3.6) Expected — Waiting for status to be reported 

I've no idea what I missed here.

**Comment by dgw** [comment]
> >  Test suite (3.6) Expected — Waiting for status to be reported 
> 
> I've no idea what I missed here.

It's because that check is marked as required in the repo settings. Before merging I would just have to update the list of expected checks.

File: .github/workflows/ci.yml
Comment (dgw) [inline .github/workflows/ci.yml:25]:
```suggestion
          - 3.9
          - 3.10
```

File: setup.cfg
Comment (dgw) [inline setup.cfg:21]:
```suggestion
    Programming Language :: Python :: 3.9
    Programming Language :: Python :: 3.10
```

File: sopel/bot.py
Comment (dgw) [inline sopel/bot.py:1161]: Surely at least `re.search` could use an intersphinx reference instead of a ha(n|r)d-coded URL? It should be possible to <tt>:ref:\`match-objects\`</tt>, too (also in `sopel/plugins/rules.py`).

Perhaps should be changed in one of the doc cleanups you/we are planning, instead of this PR, but I'm still going to call attention to it because it's being touched here. 😁

**Review by dgw — COMMENTED** [review]
Suggestions from me should be the only changes needed to add Python 3.10 to CI and reflect its "officially supported" status in the package once built.

File: sopel/bot.py
Comment (Exirel) [reply sopel/bot.py:1161]: Probably. I'll leave that to a future PR, one that fix all the "Pattern" and "Match" not found errors. 😁

**Comment by Exirel** [comment]
So Github CI thinks it's "3.1" and not "3.10"... weird.

File: .github/workflows/ci.yml
Comment (Exirel) [inline .github/workflows/ci.yml:26]:
```suggestion
          - '3.10'
```

Let's try with a quote, because apparently YAML thinks it's a float... And people wonder why I hate YAML...

**Comment by Exirel** [comment]
Alright, now we can see that Python 3.10 is not supported by `requests` yet. See also: https://pyreadiness.org/3.10/

**Comment by dgw** [comment]
I see a lot of transitive requirements in the "not supported" category, too—stuff we don't use directly, but that things we require depend on.

Let's abandon adding 3.10 in this patch and try again after the next `requests` release. (2.27 is supposed to be the last that will support Python 2.7, too, so that will be time to revisit our 7.x branch.)

**Comment by Exirel** [comment]
Weird, there is an isup test that fails in all three versions. I'm looking at potential cause, so here is the version installed:

- requests-2.27.1
- requests-mock~=1.9.3

I'll test locally to see if I can reproduce or not.

**Comment by Exirel** [comment]
I can confirm that with `request` 2.26 the test pass, not with 2.27. This will need another PR to fix that.

**Comment by Exirel** [comment]
Following [Python 3.7 documentation](https://docs.python.org/3.7/library/__future__.html), I removed the `generator_stop` and added the `annotation` future imports.

File: setup.cfg
Comment (dgw) [inline setup.cfg:56]:
```suggestion
```

We're not ignoring this one any more. Don't know if trailing comma is allowed in this value, so you might need to remove the comma after `FI17` above.

**Review by dgw — COMMENTED** [review]
One nit, for your INI-tweaking pleasure.

**Comment by Exirel** [comment]
Yup, fixed it and squashed it. Didn't change anything else.

File: sopel/bot.py
Comment (Exirel) [reply sopel/bot.py:1161]: Future PR then?

Comment (dgw) [reply sopel/bot.py:1161]: Yeah, future is fine.

**Review by dgw — APPROVED** [review]
I guess passing CI is our proof that trailing commas in flake8 config don't matter. 😁

**Pull request merged**
